# Student Performance Indicator

## Life cycle of a Machine Learning Project

* Understanding the problem statement
* Data xolletion
* Data checks to perform
* Exploratory data analysis
* Data pre-processing
* Model training
* choose best model

### 1) Problem statement

* This projects understand how the student's performance(test score) is affected by other variables such as gender, ethnicity, parental level of education, lunch and test preparation course.

### 2) Data collection

* Data source - https://www.kaggle.com/datasets/rkiattisak/student-performance-in-mathematics
* The data consits of 8 columns and 100 rows.

#### 2.1 Import data and requires packages
importing Pandas, Numpy, Matplotlib, Seaborn and Warnings library.

In [35]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')



## Import the csv Data sa Pandas DataFrame

In [36]:
import os
os.getcwd()

'd:\\GitHub\\End_To_End_ML_Projects\\src\\notebook'

In [37]:
df = pd.read_csv('data\exams_performance.csv')

In [38]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group D                some college      standard   
1    male        group D          associate's degree      standard   
2  female        group D                some college  free/reduced   
3    male        group B                some college  free/reduced   
4  female        group D          associate's degree      standard   

  test preparation course  math score  reading score  writing score  
0               completed          59             70             78  
1                    none          96             93             87  
2                    none          57             76             77  
3                    none          70             70             63  
4                    none          83             85             86

In [39]:
df.shape

(1000, 8)

### 2.2 Dataset information

* gender: sex of students->(Male/Female)
* race/ethnicity: ethnicity of students->(Group A,B,C,D,E)
* parental level of eduation: parent's final education->(bachelor's degree, some college, master's degree, associate's degree, high school)
* lunch: having lunch before test(standard or free/reduced)
* test preparation course: complete or not complete before test
* maths score
* reading score
* writing score

### 3) Data checks for perform

* check missing values
* check duplicates
* check data type
* check the number of unique values of each column
* check statistics of dataset
* check various categories present in the  different categorical column

#### 3.1 Check missing values

In [40]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

There are no missing values in the dataset

#### 3.2 Check duplicates

In [41]:
df.duplicated().sum()

np.int64(0)

There are no duplicates in the dataset

#### 3.3 Check data types

In [42]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


#### 3.4 Checking the number of unique values of each column

In [44]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     79
reading score                  71
writing score                  74
dtype: int64

#### 3.5 Check statistics of dataset

In [45]:
df.describe()

math score  reading score  writing score
count  1000.000000    1000.000000    1000.000000
mean     67.810000      70.382000      69.140000
std      15.250196      14.107413      15.025917
min      15.000000      25.000000      15.000000
25%      58.000000      61.000000      59.000000
50%      68.000000      70.500000      70.000000
75%      79.250000      80.000000      80.000000
max     100.000000     100.000000     100.000000

#### Insight

* From above description of numerical data, all means are very close to each other- between 67.8 and 70.4;
* All standard deviations are also close- between 14.10 and 15.25;
* While the minimum score for both math and writing is 15, the minimum score for reading is significantly higher at 25.

#### 3.7 Exploring Data

In [48]:
print("Categories in 'gender' variable:  ", end=" ")
print(df['gender'].unique())

print("Categories in 'race/ethnicity' variable:  ", end=" ")
print(df['race/ethnicity'].unique())

print("Categories in 'parental level of education' variable:  ", end=" ")
print(df['parental level of education'].unique())

print("Categories in 'lunch' variable:  ", end=" ")
print(df['lunch'].unique())

print("Categories in 'test preparation course' variable:  ", end=" ")
print(df['test preparation course'].unique())

Categories in 'gender' variable:   ['female' 'male']
Categories in 'race/ethnicity' variable:   ['group D' 'group B' 'group C' 'group E' 'group A']
Categories in 'parental level of education' variable:   ['some college' "associate's degree" 'some high school'
 "bachelor's degree" "master's degree" 'high school']
Categories in 'lunch' variable:   ['standard' 'free/reduced']
Categories in 'test preparation course' variable:   ['completed' 'none']
